### PR TITLE
September Funding Update Part A

### PR DESCRIPTION
**Pull request by LucasWongC**
# Changelog

This proposal updates the AaveSwapper implementation per this PR: https://github.com/bgd-labs/aave-helpers/pull/353
Here is a test swap done with this implementation:

Original tx: https://etherscan.io/tx/0x1415a8a4a3222e0a7f035496583dbcac15fecda4808261329102037a9cd25ad3
Settlement: https://etherscan.io/tx/0x3aeae7383e6b1d41715b728f0ea82db53240875329eb20556041c538cf5fface
Cow Swap UI: https://explorer.cow.fi/address/0xcd6b416C6bdF7B14C11cedcf9d61f02B28FB6fCB

### Migrate the following assets from Aave v2 to v3 Polygon.

amUSDT (All-100x10^6)
amDAI (All-1.0x10^18)
amWMATIC (All-1.0x10^18)
amWETH (All-1.0x10^18)
amWBTC (All-1.0x10^8)
amLINK (All-1.0x10^18)

### Transfer the following assets to Ethereum.

| Polygon                 | Arbitrum                 | Optimism                 |
| ----------------------- | ------------------------ | ------------------------ |
| amUSDC.e (All-100x10^6) | aArbUSDC (All-100x10^6)  | aOptUSDC (All-100x10^6)  |
| aPolUSDC (All-100x10^6) | aArbLUSD (All-1.0x10^18) | aOptLUSD (All-1.0x10^18) |

### Rescue Paraswap Funds

Rescue funds held in the Paraswap adapter contracts and send back to Treasury in line with this [PR 454](https://github.com/bgd-labs/aave-proposals-v3/pull/454) on Ethereum.

### Gas Rebate

Transfer 0.264 ETH to 0x818C277dBE886b934e60aa047250A73529E26A99 (karpatkey) reimbursing for the [funding of Guardian signers](https://governance.aave.com/t/arfc-renewal-of-aave-guardian-2024/17523/32).

### Swap funds to GHO

|        Ethereum        |
| :--------------------: |
|     aUSDC (1.25M)      |
|     aUSDT (1.25M)      |
|       DAI (All)        |
| aDAI (All-1.0x10^-18)  |
|     aEthDAI (0.5M)     |
| aLUSD (All-1.0x10^-18) |
|       LUSD (All)       |
|       FRAX (All)       |
| aFRAX (All-1.0x10^-18) |

### Merit + Ahab Programs

Create allowances to the Merit and Ahab, 3M GHO and 800 aEthWETH from Aave v3 Ethereum:
SAFE: 0xdeadD8aB03075b7FBA81864202a2f59EE25B312b

### CODE REVIEW

**Review by sakulstra — COMMENTED** [review]
in a lot of places `scaledBalanceOf` is used, but i think not in a single one it makes sense. Can you elaborate why it is used?

File: src/20241113_Multi_SeptemberFundingUpdatePartA/AaveV3Arbitrum_SeptemberFundingUpdatePartA_20241113.sol [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  // https://arbiscan.io/address/0x0335ffa9af5CE05590d6C9A75B645470e07744a9
+  IAaveArbEthERC20Bridge public constant BRIDGE =
+    IAaveArbEthERC20Bridge(0x0335ffa9af5CE05590d6C9A75B645470e07744a9);
+  // https://arbiscan.io/address/0x096760F208390250649E3e8763348E783AEF5562
+  address public constant USDC_GATEWAY = 0x096760F208390250649E3e8763348E783AEF5562;
+  // https://arbiscan.io/address/0x09e9222E96E7B4AE2a407B98d48e330053351EEe
+  address public constant LUSD_GATEWAY = 0x09e9222E96E7B4AE2a407B98d48e330053351EEe;
+
+  function execute() external {
+    _bridge();
```
[the remark this answers is not in the document]
Comment (LucasWongC) [reply src/20241113_Multi_SeptemberFundingUpdatePartA/AaveV3Arbitrum_SeptemberFundingUpdatePartA_20241113.sol:34]: It's very huge pr. and includes many changes. so I split it by sub function

File: src/20241113_Multi_SeptemberFundingUpdatePartA/AaveV3Arbitrum_SeptemberFundingUpdatePartA_20241113.sol [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * - Snapshot: Direct-to-AIP
+ * - Discussion: https://governance.aave.com/t/arfc-september-funding-update/19162
+ */
+contract AaveV3Arbitrum_SeptemberFundingUpdatePartA_20241113 is IProposalGenericExecutor {
+  using SafeERC20 for IERC20;
+  using CollectorUtils for ICollector;
+
+  // https://arbiscan.io/address/0x0335ffa9af5CE05590d6C9A75B645470e07744a9
+  IAaveArbEthERC20Bridge public constant BRIDGE =
+    IAaveArbEthERC20Bridge(0x0335ffa9af5CE05590d6C9A75B645470e07744a9);
```
[the remark this answers is not in the document]
Comment (LucasWongC) [reply src/20241113_Multi_SeptemberFundingUpdatePartA/AaveV3Arbitrum_SeptemberFundingUpdatePartA_20241113.sol:27]: okay

**Review by sakulstra — DISMISSED** [review]

File: src/20241113_Multi_SeptemberFundingUpdatePartA/AaveV3Ethereum_SeptemberFundingUpdatePartA_20241113.sol [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+        slippage: 500
+      })
+    );
+  }
+
+  function _rescueParaswap() internal {
+    IRescuable(DEBT_SWAP_ADAPTER).rescueTokens(IERC20(AaveV2EthereumAssets.sUSD_UNDERLYING));
+    IRescuable(DEBT_SWAP_ADAPTER).rescueTokens(IERC20(AaveV2EthereumAssets.USDC_UNDERLYING));
+    IRescuable(DEBT_SWAP_ADAPTER_V3).rescueTokens(IERC20(AaveV3EthereumAssets.USDT_UNDERLYING));
+    IRescuable(DEBT_SWAP_ADAPTER_V3).rescueTokens(IERC20(AaveV3EthereumAssets.crvUSD_UNDERLYING));
```
[the remark this answers is not in the document]
Comment (LucasWongC) [reply src/20241113_Multi_SeptemberFundingUpdatePartA/AaveV3Ethereum_SeptemberFundingUpdatePartA_20241113.sol:247]: okay. let me remove this

File: src/20241113_Multi_SeptemberFundingUpdatePartA/AaveV3Ethereum_SeptemberFundingUpdatePartA_20241113.sol [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+      IERC20(AaveV3EthereumAssets.rETH_UNDERLYING).balanceOf(address(this))
+    );
+    IERC20(AaveV3EthereumAssets.WBTC_UNDERLYING).transfer(
+      address(AaveV3Ethereum.COLLECTOR),
+      IERC20(AaveV3EthereumAssets.WBTC_UNDERLYING).balanceOf(address(this))
+    );
+    IERC20(AaveV3EthereumAssets.USDT_UNDERLYING).safeTransfer(
+      address(AaveV3Ethereum.COLLECTOR),
+      IERC20(AaveV3EthereumAssets.USDT_UNDERLYING).balanceOf(address(this))
+    );
```
Comment (brotherlymite) [inline src/20241113_Multi_SeptemberFundingUpdatePartA/AaveV3Ethereum_SeptemberFundingUpdatePartA_20241113.sol:277]:
```suggestion
```

File: src/20241113_Multi_SeptemberFundingUpdatePartA/AaveV3Ethereum_SeptemberFundingUpdatePartA_20241113.sol [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+      IERC20(AaveV2EthereumAssets.sUSD_UNDERLYING).balanceOf(address(this))
+    );
+    IERC20(AaveV2EthereumAssets.USDC_UNDERLYING).transfer(
+      address(AaveV3Ethereum.COLLECTOR),
+      IERC20(AaveV2EthereumAssets.USDC_UNDERLYING).balanceOf(address(this))
+    );
+    IERC20(AaveV3EthereumAssets.crvUSD_UNDERLYING).transfer(
+      address(AaveV3Ethereum.COLLECTOR),
+      IERC20(AaveV3EthereumAssets.crvUSD_UNDERLYING).balanceOf(address(this))
+    );
```
Comment (brotherlymite) [inline src/20241113_Multi_SeptemberFundingUpdatePartA/AaveV3Ethereum_SeptemberFundingUpdatePartA_20241113.sol:261]:
```suggestion
```

**Review by brotherlymite — DISMISSED** [review]

**Review by sakulstra — APPROVED** [review]

**Pull request merged**